### PR TITLE
[FEATURE] Instance Branching

### PR DESCRIPTION
**Pull request by pkitazos**
**Completed**

- Admin can fork an allocation instance after completing a round of matching to deal with any unmatched students
- Admin can merge a forked allocation instance back into its parent instance after completing an nth round of matching in said forked instance
- Added info notes to Instance edit form to inform admin of potentially unclear merging behaviour when making changes to an Instance's flags and tags
- Added info notes to Project edit form to inform admin or supervisor of potentially unclear merging behaviour when making changes to a projects flags and tags
- Created utility functions for identifying items by fields other than their primary identifier but which are also unique (title for projects, userId for project allocations)
- Minor UI fixes for consistency in manual changes admin panel page
- Moved admin panel tab computation to server and created config file containing all paths and display names

### CODE REVIEW

**Comment by AlysonDick** [comment]
**FIX REQUIRED** Students "My preferences" do not updated after merge

Incorrect information is displayed on a student's My Preferences page. This occurs when a student has submitted preferences in the parent instance and then submits new preferences in the forked instance.

**Comment by AlysonDick** [comment]
**FIX REQUIRED** Student not being allocated correctly to a project in forked instance that has a different student allocated in parent instance.

Bug found when testing upper bound capacities of projects. To recreate error:

1. Login as admin and change stage to 2 
2. Edit the capacity upper bound of project "Chatbot Assistant" to >1
3. Change stage to 4 and run and select "Generous" algorithm
4. Change to stage 5, move to manual adjustment page and select all students
5. Check that one student is assigned to project 001 (student 234576n should be assigned)
6. Unallocate students on projects 3 and 4 (students 234578a. and 2345679k, they should be 3rd and 4th from bottom of table) - note, I did this to ensure the supervisor would have capacity for more students
7. Submit the changes
8. Change to stage 6 and fork (note, three students, two supervisors and six projects will be carried to forked instance)
9. Change forked instance stage to 3
10. Edit the instance - change the minimum num. of preferences to 1 under student restrictions
11. Login as a student, add the project "Chatbot Assistant" as a preference and then submit preferences
12. As admin, move stage to 4 and run and select algorithms as normal - Screenshot 1 below looks as though the student was allocated successfully
13. View algorithm details - the student evaluator does not show here, but one of the other students carried over does?
14. Move to stage 6 and merge
15. Check the allocation overview after merging and see that the allocation has not been made


Side note: if the student is manually allocated then the allocation will show in the allocation overview on the parent instance after merging BUT this causing unusual behaviour... First, the value in student rank column of the table on the parent instance on the allocation overview page is reduced by 1. Secondly, if you try to view "My allocations" in the forked instance a runtime error is displayed on screen.

**Comment by pkitazos** [comment]
> **FIX REQUIRED** Students "My preferences" do not updated after merge
> 
> Incorrect information is displayed on a student's My Preferences page. This occurs when a student has submitted preferences in the parent instance and then submits new preferences in the forked instance.

So the expected behaviour would be that the parent instance preferences would be overwritten by the forked instance preferences right?

**Comment by pkitazos** [comment]
> **FIX REQUIRED** Student not being allocated correctly to a project in forked instance that has a different student allocated in parent instance. Unsure if related to other known issue.
> 
> Bug found when testing upper bound capacities of projects. To recreate error:
> 
> 1. Login as admin and change stage to 2
> 2. Edit the capacity upper bound of project "Chatbot Assistant" to >1
> 3. Change stage to 4 and run and select "Generous" algorithm
> 4. Change to stage 5, move to manual adjustment page and select all students
> 5. Check that one student is assigned to project 001 (student 234576n should be assigned)
> 6. Unallocate students on projects 3 and 4 (students 234578a. and 2345679k, they should be 3rd and 4th from bottom of table) - note, I did this to ensure the supervisor would have capacity for more students
> 7. Submit the changes
> 8. Change to stage 6 and fork (note, three students, two supervisors and six projects will be carried to forked instance)
> 9. Change forked instance stage to 3
> 10. Edit the instance - change the minimum num. of preferences to 1 under student restrictions
> 11. Login as a student, add the project "Chatbot Assistant" as a preference and then submit preferences
> 12. As admin, move stage to 4 and run and select algorithms as normal - Screenshot 1 below looks as though the student was allocated successfully
> 13. View algorithm details - the student evaluator does not show here, but one of the other students carried over does?
> 14. Move to stage 6 and merge
> 15. Check the allocation overview after merging and see that the allocation has not been made
> 
> Side note: if the student is manually allocated then the allocation will show in the allocation overview on the parent instance after merging BUT this causing unusual behaviour... First, the value in student rank column of the table on the parent instance on the allocation overview page is reduced by 1. Secondly, if you try to view "My allocations" in the forked instance a runtime error is displayed on screen.

@AlysonDick Looks like you may have forgotten to attach the screenshot, in any case, I'm not able to recreate this bug. I get to point 12 and it appears as though the student evaluator has been successfully allocated that project. This persists through 13-15 and when I merge the instance it seems like the student evaluator has in fact been allocated to the project. I think perhaps the fix for the previously known issue, has handled this situation too. We should probably test it a bit more thoroughly after discussing the issue with David.

**Review by pkitazos — COMMENTED** [review]
Looks good, let's get this merged in!

File: docs/merge-plan.txt
[the remark this answers is not in the document]
Comment (AlysonDick) [reply docs/merge-plan.txt:1]: Plan to be kept for now. Conversion to documentation required in future.

File: src/server/routers/institution/instance/_utils/merge.ts
[the remark this answers is not in the document]
Comment (AlysonDick) [reply src/server/routers/institution/instance/_utils/merge.ts:1]: Comments now removed.

**Review by AlysonDick — APPROVED** [review]
Ready to be merged.

**Pull request merged**